### PR TITLE
refactor(rome_formatter): Use custom trait for unique `LabelId`s

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

See https://github.com/astral-sh/ruff/issues/5263

Rust 1.72.0 changes `TypeId` from a `u64` to a `u128`. This breaks Rome's release build because it increases the size of `FormatElement`.

This PR removes the dependency on `TypeId` for `LabelId` and instead introduces a new trait that downstream crates should implement on an `enum` defining all labels for that project. 

This approach has two downsides:

* The API cannot statically enforce that a project uses a single `Label`. Mixing two `Label` could result in two labels being equal that should not be equal. I added a 
debug assertion that compares the names of the labels to mitigate the risk. 
* Implementing the first `Label` requires more boilerplate code than before

One main upside is that it allows a project to define all labels in a single place (and search for all label usages).  

## Test Plan

`cargo test` and `cargo build --release -p rome_formatter` with Rust 1.72.0

## Changelog

- [ ] The PR requires a changelog line

## Documentation

- [ ] The PR requires documentation
- [ ] I will create a new PR to update the documentation

### CODE REVIEW

File: crates/rome_js_formatter/src/lib.rs [hunk 1]
```diff
@@ -196,143 +197,7 @@ use crate::context::{JsFormatContext, JsFormatOptions};
 use crate::cst::FormatJsSyntaxNode;
 use crate::syntax_rewriter::transform;
 
-/// Used to get an object that knows how to format this object.
```
Comment (MichaReiser) [inline crates/rome_js_formatter/src/lib.rs:199]: This is unrelated to the change. Happy to undo. But we should either remove the `shared_traits.rs` file or include it.

Comment (ematipico) [reply crates/rome_js_formatter/src/lib.rs:199]: Probably, we want to remove the file and keep the code.

This code would file when publishing the crate, **I think**.

**Review by ematipico — APPROVED** [review]
Thank you @MichaReiser !! 🙏

File: crates/rome_js_formatter/src/lib.rs (hunk 1 above)
Comment (MichaReiser) [reply crates/rome_js_formatter/src/lib.rs:199]: Okay. I'll undo
